sklearn/utils/stats.py

```diff
@@ -71,60 +63,38 @@ def _weighted_percentile(array, sample_weight, percentile=50,
     def _squeeze_arr(arr, n_dim):
         return arr[0] if n_dim == 1 else arr
 
-    if interpolation == "nearest":
-        # compute by nearest-rank method
-        # The rank correspond to P / 100 * N; P in [0, 100]
-        # Here, N is replaced by the sum of the weight and P will be adjusted
-        # by the weights
-        adjusted_percentile = percentile * weight_cdf[-1]
-        percentile_value_idx = np.array([
-            np.searchsorted(weight_cdf[:, i], adjusted_percentile[i])
-            for i in range(weight_cdf.shape[1])
+    adjusted_percentile = (weight_cdf - sorted_weights)
+    with np.errstate(invalid="ignore"):
+        adjusted_percentile /= ((weight_cdf[-1] * (n_rows - 1)) / (n_rows))
```

Comment:
I don't quite understand why `((weight_cdf[-1] * (n_rows - 1)) / (n_rows))` ?

Reply:
With unit weight you have

```
p = (x - 1) / (N - 1)
```

Taking the weight into account, it comes back to

```
p = (x - w_n) / (S_n - (S_n / N_non_zero))
```

Reply:
`n_rows` was wrong because it was not discarding `zero` weights elements (which was the issue in `neg_median_score` tests.

Reply:
With `p = (x - 1) / (N - 1)`, for 1,2,3,4,5 and `x=2` and `N=5`, p = 1/4. The purpose of the '-1' is to remove the `x` element from counting, so we calculate that 1/4 of elements are less than x and 3/4 of elements larger than x.

For weighted shouldn't it be:
```
p = x_cumsum[i] - x[i] / x_cumsum[-1] - x[i]
```
where the `i` is the element in question...?

Reply:
I think `N` can be thought of as cumsum when weights are 1.
Thus for weighted percentile, I think it shouldn't be count (`non_zero = np.count_nonzero(sorted_weights, axis=0)`) that you're dividing by, but the cumsum (minus the weight of the element in question)

Reply:
Actually I'm not sure if this translates simply to weighted percentile. Maybe this is most suitable: https://en.wikipedia.org/wiki/Percentile#The_weighted_percentile_method

Reply:
This definition is an extension of the first approach. I assume that you extend the 2nd and 3rd variants. And the second variant is the one used by NumPy.

Reply:
> Thus for weighted percentile, I think it shouldn't be count (non_zero = np.count_nonzero(sorted_weights, axis=0)) that you're dividing by, but the cumsum (minus the weight of the element in question)

I think that you are right.

Reply:
Then we also need better tests for the weighted percentile because we should catch this case.

Reply:
I think that the issue was that our tests are to close to unit weights and the normalization was wrongly working.

Reply:
OK I added a test with some random weights which was failing with the bug.